### PR TITLE
Updates and Improvements to Unit Tests

### LINKED ISSUE #290: [Feature Contribution]: Improve unit test coverage and style
### What happened?

As a Carbon Aware application developer, I want to know that all the existing codepaths have been unit tested to ensure bugs are caught.

# Acceptance Criteria

- [x] Ensure that all external client functions (WebAPi, CLI, and Library) are fully unit tested and code coverage confirms (ignoring non-testable code paths). Example of `forecast/batch` not having tests.
- [x] Ensure that the main aggregator and data source functions (which are part of a flow from the client) are fully unit tested and code coverage confirms (ignoring non-testable code paths)
- [x] Condense redundantly similar tests into single test with multiple test cases (exception throwing, null/edge cases are good ones usually)
- [x] Ensure complex functions have edge case testing too, not just happy path

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Feature Commitment

- [X] I commit to contributing this feature as a PR and working with the GSF to merge this feature into the Carbon Aware SDK.

### PR DESCRIPTION
**Pull request by Kpopovs1**
# Pull Request

Issue Number: #290

## Summary

As a Carbon Aware application developer, I want to know that all the existing codepaths have been unit tested to ensure bugs are caught.

## Changes

- Reviewed unit tests throughout the project and added new tests to improve code coverage.

## Checklist

- [ ] Local Tests Passing?
- [ ] CICD and Pipeline Tests Passing?
- [ ] Added any new Tests?
- [ ] Documentation Updates Made?
- [ ] Are there any API Changes? If yes, please describe below.
- [ ] This is not a breaking change. If it is, please describe it below.

## Are there API Changes?

No.

## Is this a breaking change?

No.

> Please follow
> [GitHub's suggested syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
> to link Pull Requests to Issues via keywords

This PR Closes #290

### CODE REVIEW

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Configuration/ServiceCollectionExtensionTests.cs [hunk 1]
```diff
@@ -35,5 +38,61 @@ public void ClientProxyTest_With_Missing_ProxyURL_ThrowsException()
 
         // Act & Assert
         Assert.Throws<ConfigurationException>(() => serviceCollection.AddElectricityMapsForecastDataSource(configuration.DataSources()));
+        Assert.Throws<ConfigurationException>(() => serviceCollection.AddElectricityMapsEmissionsDataSource(configuration.DataSources()));
+    }
+
+    [Test]
```
Comment (juzuluag) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Configuration/ServiceCollectionExtensionTests.cs:44]: Suggestion: `ClientProxyTest_With_No_ProxyURL_AddsServices ` and `ClientProxyTest_With_ProxyURL_AddsServices` seem like they can be consolidated into one `TestCase` scenarios, since the code looks pretty similar. Passing a boolean flag can help to create a single test method with 2 test cases.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Configuration/ServiceCollectionExtensionTests.cs [hunk 2]
```diff
@@ -35,5 +38,61 @@ public void ClientProxyTest_With_Missing_ProxyURL_ThrowsException()
...
+            .AddEnvironmentVariables()
+            .Build();
+        var serviceCollection = new ServiceCollection();
+
+        // Act
+        var forecastService = serviceCollection.AddElectricityMapsForecastDataSource(configuration.DataSources());
+        var emissionsService = serviceCollection.AddElectricityMapsEmissionsDataSource(configuration.DataSources());
+
+        // Assert
+        Assert.NotNull(forecastService);
```
Comment (juzuluag) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Configuration/ServiceCollectionExtensionTests.cs:66]: nit: change for `Assert.That(...)` approach
```suggestion
 Assert.That(forecastService, Is.Not.Null);
```

File: src/CarbonAware.DataSources/CarbonAware.DataSources.Json/test/JsonDataSourceTests.cs [hunk 3]
```diff
@@ -82,6 +82,29 @@ public async Task GetCarbonIntensityAsync_ReturnsSingleDataPoint_WhenStartParamE
...
+            .ReturnsAsync(new List<EmissionsData>())
+            .Verifiable();
+
+        var location = new Location() { Name = "midwest" };
+        var locations = new List<Location>() { location };
+        var start = DateTimeOffset.Parse("2022-09-07T12:45:11+00:00");
+        var end = DateTimeOffset.Parse("2022-09-07T13:45:11+00:00");
+        var dataSource = mockDataSource.Object;
+        var result = await dataSource.GetCarbonIntensityAsync(locations, start, end);
+        Assert.AreEqual(0, result.Count());
```
Comment (juzuluag) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.Json/test/JsonDataSourceTests.cs:104]: Similar comment with regards `Assert.That()`

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/Configuration/ServiceCollectionExtensionTests.cs [hunk 4]
```diff
@@ -66,5 +70,33 @@ public void ClientProxyTest_With_Missing_ProxyURL_ThrowsException()
 
         // Act & Assert
         Assert.Throws<ConfigurationException>(() => serviceCollection.AddWattTimeForecastDataSource(configuration.DataSources()));
+        Assert.Throws<ConfigurationException>(() => serviceCollection.AddWattTimeEmissionsDataSource(configuration.DataSources()));
+    }
+
+    [Test]
+    public void ClientProxyTest_With_Missing_ProxyURL_AddsDataSource()
```
Comment (juzuluag) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/Configuration/ServiceCollectionExtensionTests.cs:77]: Same comments as the one for Electricity Map. Do you know if there is a need for the None Proxy Case?

File: src/CarbonAware.LocationSources/test/LocationSourceTest.cs [hunk 5]
```diff
@@ -231,6 +230,28 @@ public async Task GeopositionLocation_ValidLocation_DupLocationKey()
...
+            Prefix = "prefix",
+            Delimiter = "-",
+            DataFileLocation = _goodFile
+        });
+        var options = new Mock<IOptionsMonitor<LocationDataSourcesConfiguration>>();
+        options.Setup(o => o.CurrentValue).Returns(() => configuration);
+        var logger = Mock.Of<ILogger<LocationSource>>();
+        var locationSource = new LocationSource(logger, options.Object);
+
+        IDictionary<string, Location> allLocations = await locationSource.GetGeopositionLocationsAsync();
```
Comment (juzuluag) [inline src/CarbonAware.LocationSources/test/LocationSourceTest.cs:248]: nit:
```suggestion
var allLocations = await locationSource.GetGeopositionLocationsAsync();
```

File: src/CarbonAware.LocationSources/test/LocationSourceTest.cs [hunk 6]
```diff
@@ -231,6 +230,28 @@ public async Task GeopositionLocation_ValidLocation_DupLocationKey()
...
+
+        Assert.AreEqual(allLocations.Count, 3);
```
Comment (juzuluag) [inline src/CarbonAware.LocationSources/test/LocationSourceTest.cs:250]: Same comment about `Assert.That`

File: src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        // Act & Assert
+        Assert.DoesNotThrow(() => ServiceCollectionExtensions.AddMonitoringAndTelemetry(services, configuration));
+        Assert.AreEqual(services.Count(), 43);
+    }
+
+    [Test]
+    public void AddMonitoringAndTelemetry_AddsServices_WithInstrumentationKey()
+    {
+        // Arrange
+        IServiceCollection services = new ServiceCollection();
```
Comment (juzuluag) [inline src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs:36]: nit:
```suggestion
var services = new ServiceCollection();
```

File: src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+        var inMemorySettings = new Dictionary<string, string>
+        {
+            { "CarbonAwareVars:TelemetryProvider", "ApplicationInsights" },
+            { "AppInsights_InstrumentationKey", "AppInsightsInstrumentationKey" }
+        };
+        IConfiguration configuration = new ConfigurationBuilder()
```
Comment (juzuluag) [inline src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs:43]: Same here `var ...`

File: src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            .AddInMemoryCollection(inMemorySettings)
+            .Build();
+
+        // Act & Assert
+        Assert.DoesNotThrow(() => ServiceCollectionExtensions.AddMonitoringAndTelemetry(services, configuration));
+        Assert.AreEqual(services.Count(), 43);
```
Comment (juzuluag) [inline src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs:49]: nit:
`Assert.That`

File: src/CarbonAware.WebApi/test/unitTests/Filters/CarbonAwareParametersBaseDtoSchemaFilterTests.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+using NUnit.Framework;
+using Swashbuckle.AspNetCore.SwaggerGen;
+using GSF.CarbonAware.Handlers.CarbonAware;
+using System.Text.Json;
+
+namespace CarbonAware.WepApi.UnitTests;
+
+[TestFixture]
+
+public class CarbonAwareParametersBaseDtoSchemaFilterTests
```
Comment (juzuluag) [inline src/CarbonAware.WebApi/test/unitTests/Filters/CarbonAwareParametersBaseDtoSchemaFilterTests.cs:12]: Maybe we can take this offline and you elaborate more,  since I'm not sure what the utest is about.

File: src/CarbonAware/test/Configuration/DataSourcesConfigurationTests.cs [hunk 11]
```diff
@@ -57,4 +57,37 @@ public void AssertValid_DataSourceNotInConfigurationSection_ThrowsException(stri
         Assert.That(ex!.Message, Contains.Substring(errorMessage));
     }
 
+    [Test]
```
Comment (juzuluag) [inline src/CarbonAware/test/Configuration/DataSourcesConfigurationTests.cs:60]: This is great test. It feels like this is another candidate to be a `TestCase` method.

File: src/GSF.CarbonAware/test/Configuration/ServiceCollectionExtensionsTests.cs [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+namespace GSF.CarbonAware.Tests;
+
+[TestFixture]
+public class ServiceCollectionExtensionsTest
+{
+    [Test]
+    public void AddEmissionsServices_ReturnsServices()
+    {
+        // Arrange
+        IServiceCollection services = new ServiceCollection();
```
Comment (juzuluag) [inline src/GSF.CarbonAware/test/Configuration/ServiceCollectionExtensionsTests.cs:17]: nit: 
`var ...`

File: src/GSF.CarbonAware/test/Configuration/ServiceCollectionExtensionsTests.cs [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+        };
+        IConfiguration configuration = new ConfigurationBuilder()
+            .AddInMemoryCollection(inMemorySettings)
+            .Build();
+
+        // Act
+        IServiceCollection provider = ServiceCollectionExtensions.AddEmissionsServices(services, configuration);
+
+        // Assert
+        Assert.NotNull(provider);
```
Comment (juzuluag) [inline src/GSF.CarbonAware/test/Configuration/ServiceCollectionExtensionsTests.cs:37]: nit: `Assert.That()`

**Review by juzuluag — CHANGES_REQUESTED** [review]
This is looking good. Added a couple of comments and nits. Thanks for doing this @Kpopovs1 , @in4margaret and @JenMadiedo

File: src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+namespace CarbonAware.WepApi.UnitTests;
+
+[TestFixture]
+public class ServiceCollectionExtensionsTests
+{
+    [Test]
+    public void AddMonitoringAndTelemetry_AddsServices_WithConnectionString()
+    {
+        // Arrange
+        IServiceCollection services = new ServiceCollection();
```
Comment (juzuluag) [inline src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs:16]: nit: `var ...`

File: src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+        {
+            { "CarbonAwareVars:TelemetryProvider", "ApplicationInsights" },
+            { "ApplicationInsights_Connection_String", "AppInsightsConnectionString" }
+        };
+        IConfiguration configuration = new ConfigurationBuilder()
+            .AddInMemoryCollection(inMemorySettings)
+            .Build();
+
+        // Act & Assert
+        Assert.DoesNotThrow(() => ServiceCollectionExtensions.AddMonitoringAndTelemetry(services, configuration));
```
Comment (juzuluag) [inline src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs:28]:
```suggestion
        Assert.DoesNotThrow(() => services.AddMonitoringAndTelemetry(configuration));
```

File: src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+        Assert.AreEqual(services.Count(), 43);
```
Comment (juzuluag) [inline src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs:29]: QQ: How do you know this exact number?

File: src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+        {
+            { "CarbonAwareVars:TelemetryProvider", "ApplicationInsights" },
+            { "AppInsights_InstrumentationKey", "AppInsightsInstrumentationKey" }
+        };
+        IConfiguration configuration = new ConfigurationBuilder()
+            .AddInMemoryCollection(inMemorySettings)
+            .Build();
+
+        // Act & Assert
+        Assert.DoesNotThrow(() => ServiceCollectionExtensions.AddMonitoringAndTelemetry(services, configuration));
```
Comment (juzuluag) [inline src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs:48]:
```suggestion
        Assert.DoesNotThrow(() => services.AddMonitoringAndTelemetry(configuration));
```

File: src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs (hunk 9 above)
Comment (juzuluag) [inline src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs:49]: Magic number (43)? :-)

File: src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+        var inMemorySettings = new Dictionary<string, string>
+        {
+            { "CarbonAwareVars:TelemetryProvider", "ApplicationInsights" }
+        };
+        IConfiguration configuration = new ConfigurationBuilder()
+            .AddInMemoryCollection(inMemorySettings)
+            .Build();
+
+        // Act & Assert
+        Assert.DoesNotThrow(() => ServiceCollectionExtensions.AddMonitoringAndTelemetry(services, configuration));
```
Comment (juzuluag) [inline src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs:67]:
```suggestion
        Assert.DoesNotThrow(() => services.AddMonitoringAndTelemetry(configuration));
```

File: src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+        // Arrange
+        IServiceCollection services = new ServiceCollection();
+
+        var inMemorySettings = new Dictionary<string, string>{};
+        IConfiguration configuration = new ConfigurationBuilder()
+            .AddInMemoryCollection(inMemorySettings)
+            .Build();
+
+        // Act & Assert
+        Assert.DoesNotThrow(() => ServiceCollectionExtensions.AddMonitoringAndTelemetry(services, configuration));
```
Comment (juzuluag) [inline src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs:83]:
```suggestion
        Assert.DoesNotThrow(() => services.AddMonitoringAndTelemetry(configuration));
```

**Review by juzuluag — CHANGES_REQUESTED** [review]
More comments.

File: src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs (hunk 16 above)
Comment (Kpopovs1) [reply src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs:29]: I ran `AddMonitoringAndTelemetry` and checked the number of services registered to the collection when it runs successfully. I added it to contrast that when this is not successful, the number of services registered is `0` in the other tests.

Thinking about it in depth, we don't need to add how many services the collection has here as long as the other tests demonstrate the number is `0` when the call throws an error (also this `43` may change in the future lol). I removed this assertion statement and the one below, let me know if that looks good!

File: src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs (hunk 9 above)
Comment (Kpopovs1) [reply src/CarbonAware.WebApi/test/unitTests/Configuration/ServiceCollectionExtensionsTests.cs:49]: same comment as above!

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/Configuration/ServiceCollectionExtensionTests.cs (hunk 4 above)
Comment (Kpopovs1) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/test/Configuration/ServiceCollectionExtensionTests.cs:77]: Mainly to increase code coverage for the files (both of them), since this path (use proxy = false) showed up as untested on my code analyzer locally.

File: src/CarbonAware.WebApi/test/unitTests/Filters/CarbonAwareParametersBaseDtoSchemaFilterTests.cs (hunk 10 above)
Comment (Kpopovs1) [reply src/CarbonAware.WebApi/test/unitTests/Filters/CarbonAwareParametersBaseDtoSchemaFilterTests.cs:12]: Will ping you to discuss!

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Configuration/ServiceCollectionExtensionTests.cs [hunk 20]
```diff
@@ -43,7 +46,7 @@ public void ClientProxyTest_With_Missing_ProxyURL_ThrowsException()
 
     [TestCase(true, TestName = "ClientProxyTest, successful: denotes adding ElectricityMaps data sources using proxy.")]
     [TestCase(false, TestName = "ClientProxyTest, successful: denotes adding ElectricityMaps data sources without using proxy.")]
-    public void ClientProxyTest_AddsServices(bool withProxyUrl)
+    public void ClientProxyTest_AddsDataSource(bool withProxyUrl)
```
Comment (juzuluag) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Configuration/ServiceCollectionExtensionTests.cs:49]: It feels like the method name should be something different, like: `Client_Proxy_Configuration()`, since testing with a proxy in the middle, it requires to perform a remote call by the client.

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Configuration/ServiceCollectionExtensionTests.cs [hunk 21]
```diff
@@ -56,9 +59,9 @@ public void ClientProxyTest_AddsServices(bool withProxyUrl)
 
         if (withProxyUrl)
         {
-            settings.Add("DataSources:Configurations:ElectricityMapsTest:Proxy:url", "http://10.10.10.1");
-            settings.Add("DataSources:Configurations:ElectricityMapsTest:Proxy:username", "proxyUsername");
-            settings.Add("DataSources:Configurations:ElectricityMapsTest:Proxy:password", "proxyPassword");
+            settings.Add(ProxyUrl, "http://10.10.10.1");
```
Comment (juzuluag) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Configuration/ServiceCollectionExtensionTests.cs:62]:
```suggestion
settings.Add(ProxyUrl, "http://fakeProxy");
```

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Configuration/ServiceCollectionExtensionTests.cs [hunk 22]
```diff
@@ -18,6 +18,9 @@ public class ServiceCollectionExtensionTests
     private readonly string TokenKey = $"DataSources:Configurations:ElectricityMapsTest:APIToken";
     private readonly string DefaultTokenValue = "myDefaultToken123";
     private readonly string UseProxyKey = $"DataSources:Configurations:ElectricityMapsTest:Proxy:UseProxy";
+    private readonly string ProxyUrl = $"DataSources:Configurations:ElectricityMapsTest:Proxy:Url";
```
Comment (juzuluag) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Configuration/ServiceCollectionExtensionTests.cs:21]: nit: Convert these `private readonly string` to `const string`. These are not initialized in the constructor,

**Review by juzuluag — COMMENTED** [review]
Few comments. Looking good 👍

File: src/CarbonAware/test/Configuration/DataSourcesConfigurationTests.cs [hunk 23]
```diff
@@ -57,37 +57,40 @@ public void AssertValid_DataSourceNotInConfigurationSection_ThrowsException(stri
...
+    [TestCase("WattTime", "ElectricityMaps", TestName = "Assert Configuration Type is forecast=WattTime and emissions=ElectricityMaps")]
+    [TestCase("Json", null, TestName = "Assert Configuration Type is forecast=Json and emissions=null")]
+    public void ConfigurationTypesAndSections_ReturnCorrectType(string forecastDataSource, string emissionsDataSource)
     {
         // Arrange
         var inMemorySettings = new Dictionary<string, string>()
         {
             { "Configurations:WattTime:Type", "WattTime" },
-            { "Configurations:Json:Type", "Json" }
+            { "Configurations:Json:Type", "Json" },
```
Comment (juzuluag) [inline src/CarbonAware/test/Configuration/DataSourcesConfigurationTests.cs:69]: Should be these configuration params also use what is passed through the parameters of the method? `forecastDataSource` and `emissionsDataSource`?
```c#
public void ConfigurationTypesAndSections_ReturnCorrectType(string confType)
{
...
{ $"Configuration:{confType}:Type", $"{confType}"},
...
}
```

File: src/CarbonAware.WebApi/test/unitTests/Controllers/CarbonAwareControllerTests.cs [hunk 24]
```diff
@@ -243,4 +244,49 @@ public async Task GetEmissionsDataForLocationsByTime_NoLocations_ThrowsException
...
+            emissionsForecastBatch.SingleLocation,
+            emissionsForecastBatch.Start,
+            emissionsForecastBatch.End,
+            emissionsForecastBatch.Requested,
+            emissionsForecastBatch.Duration)).ReturnsAsync(emissionsForecast);
+
+        // Act
+        IActionResult result = await controller.BatchForecastDataAsync(new List<EmissionsForecastBatchParametersDTO>() { emissionsForecastBatch });
+
+        //Assert
```
Comment (pritipath) [inline src/CarbonAware.WebApi/test/unitTests/Controllers/CarbonAwareControllerTests.cs:289]: Should we also verify that appropriate method in the handler is called once (like we do for other tests)?

**Review by pritipath — APPROVED** [review]
Thanks for improving the test coverage. Left a small nit comment, otherwise looks great

File: src/CarbonAware/test/Configuration/DataSourcesConfigurationTests.cs (hunk 23 above)
Comment (Kpopovs1) [reply src/CarbonAware/test/Configuration/DataSourcesConfigurationTests.cs:69]: We could do that, but we don't necessarily need to. I decided to include all three types and remove the need to handle the case where a confType for one of the data sources if null from the testcase. Let me know if you see a better case for doing it this way though!

File: src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Configuration/ServiceCollectionExtensionTests.cs (hunk 22 above)
Comment (Kpopovs1) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Configuration/ServiceCollectionExtensionTests.cs]: Done, that makes sense!

File: src/CarbonAware.WebApi/test/unitTests/Controllers/CarbonAwareControllerTests.cs (hunk 24 above)
Comment (Kpopovs1) [reply src/CarbonAware.WebApi/test/unitTests/Controllers/CarbonAwareControllerTests.cs:289]: Good catch! Done 👍

**Review by juzuluag — APPROVED** [review]
Great work!. 👍

**Comment by Kpopovs1** [comment]
This is ready for review @vaughanknight , @gfmatthews

**Review by vaughanknight — APPROVED** [review]
Approved.

**Review by Willmish — CHANGES_REQUESTED** [review]
See Merge conflicts after merging other PRs

**Comment by Willmish** [comment]
Hi @Kpopovs1 - I can see that you are struggling with getting the DCO check to pass - its a new requirement in all GSF repositories. Please see #309  for more info. 

As you are the sole author of these commits, you can rebase them and add signoff to each one of them. However, I can see that you are using the preferred way of adding a remediation commit. Since it looks like you have sign-offs from two seperate github emails, might be easier for you to reset the remediation commits and force push a clean remediation commit with the proper signoffs.

**Review by vaughanknight — APPROVED** [review]
Second round of reviews, looks good.
